mobile/library/common/http/client.cc

```diff
@@ -406,54 +412,77 @@ void Client::DirectStream::saveLatestStreamIntel() {
+  if (error_.has_value()) {
+  }
   envoy_error error{};
+  error_ = error;
+
   OptRef<RequestDecoder> request_decoder = direct_stream_.requestDecoder();
   if (!request_decoder) {
-    return error;
+    error_->message = envoy_nodata;
+    return;
   }
   const auto& info = request_decoder->streamInfo();
 
   if (info.responseCode().has_value()) {
-    error.error_code = Bridge::Utility::errorCodeFromLocalStatus(
+    error_->error_code = Bridge::Utility::errorCodeFromLocalStatus(
         static_cast<Http::Code>(info.responseCode().value()));
   } else if (StreamInfo::isStreamIdleTimeout(info)) {
-    error.error_code = ENVOY_REQUEST_TIMEOUT;
+    error_->error_code = ENVOY_REQUEST_TIMEOUT;
   } else {
-    error.error_code = ENVOY_STREAM_RESET;
+    error_->error_code = ENVOY_STREAM_RESET;
   }
 
   if (info.responseCodeDetails().has_value()) {
-    error.message = Data::Utility::copyToBridgeData(info.responseCodeDetails().value());
+    error_->message = Data::Utility::copyToBridgeData(info.responseCodeDetails().value());
   } else {
-    error.message = envoy_nodata;
+    error_->message = envoy_nodata;
   }
 
-  error.attempt_count = info.attemptCount().value_or(0);
-  return error;
+  error_->attempt_count = info.attemptCount().value_or(0);
 }
 
 Client::DirectStream::DirectStream(envoy_stream_t stream_handle, Client& http_client)
     : stream_handle_(stream_handle), parent_(http_client) {}
 
 Client::DirectStream::~DirectStream() { ENVOY_LOG(debug, "[S{}] destroy stream", stream_handle_); }
 
+CodecEventCallbacks*
+Client::DirectStream::registerCodecEventCallbacks(CodecEventCallbacks* codec_callbacks) {
+  // registerCodecEventCallbacks is called with nullptr when the underlying
+  // Envoy stream is going away. Make sure Envoy Mobile sees the stream as
+  // closed as well.
+  if (codec_callbacks == nullptr &&
+      parent_.getStream(stream_handle_, ALLOW_ONLY_FOR_OPEN_STREAMS)) {
+    // Generally this only happens if Envoy closes the (virtual) downstream
+    // connection, otherwise Envoy would inform the codec to send a reset.
+    callbacks_->closeStream(false);
+  }
+  std::swap(codec_callbacks, codec_callbacks_);
+  return codec_callbacks;
+}
+
 // Correctly receiving resetStream() for errors in Http::Client relies on at least one filter
 // resetting the stream when handling a pending local response. By default, the LocalReplyFilter
 // has this responsibility.
 void Client::DirectStream::resetStream(StreamResetReason reason) {
   // This seems in line with other codec implementations, and so the assumption is that this is in
   // line with upstream expectations.
   // TODO(goaway): explore an upstream fix to get the HCM to clean up ActiveStream itself.
-  saveFinalStreamIntel(); // Take a snapshot now in case the stream gets destroyed.
+  saveFinalStreamIntel();   // Take a snapshot now in case the stream gets destroyed.
+  callbacks_->latchError(); // Latch the error in case the stream gets destroyed.
```

Comment:
this is moved earlier because once we runResetCallbacks, we can't latch the error as it triggers the envoy-bug "don't snag stream info after we're on the deferred delete list"